src/task/SendAndReceiveDataMessagesTask.ts

```diff
@@ -0,0 +1,100 @@
+import SignalingClientEvent from '../signalingclient/SignalingClientEvent';
+import SignalingClientEventType from '../signalingclient/SignalingClientEventType';
+import SignalingClientObserver from '../signalingclientobserver/SignalingClientObserver';
+import {
+  SdkDataMessageFrame,
+  SdkDataMessagePayload,
+  SdkSignalFrame,
+} from '../signalingprotocol/SignalingProtocol.js';
+import BaseTask from './BaseTask';
+
+export default class SendAndReceiveDataMessagesTask extends BaseTask
+  implements RemovableObserver, SignalingClientObserver {
+  protected taskName = 'SendAndReceiveDataMessagesTask';
+
+  private static TOPIC_REGEX = new RegExp(/^[a-zA-Z0-9_-]{1,36}$/);
+  private static DATA_SIZE = 2048;
+
+  constructor(private context: AudioVideoControllerState) {
+    super(context.logger);
+  }
+
+  async run(): Promise<void> {
+    this.context.removableObservers.push(this);
+    this.context.signalingClient.registerObserver(this);
+    this.context.realtimeController.realtimeSubscribeToSendDataMessage(this.sendDataMessageHandler);
+  }
+
+  removeObserver(): void {
+    this.context.realtimeController.realtimeUnsubscribeFromSendDataMessage(
+      this.sendDataMessageHandler
+    );
+    this.context.signalingClient.removeObserver(this);
+  }
+
+  handleSignalingClientEvent(event: SignalingClientEvent): void {
+    if (
+      event.type === SignalingClientEventType.ReceivedSignalFrame &&
+      event.message.type === SdkSignalFrame.Type.DATA_MESSAGE
+    ) {
+      for (let message of event.message.dataMessage.dataMessagePayloads) {
+        const dataMessage = new DataMessage(
+          (message.ingestTimeNs as number) / 1000000,
+          message.topic,
+          message.data,
+          message.senderAttendeeId,
+          message.senderExternalUserId,
+          (message.ingestTimeNs as number) === 0
+        );
+        this.context.realtimeController.realtimeReceiveDataMessage(dataMessage);
+      }
+    }
+  }
+
+  sendDataMessageHandler = (
+    topic: string,
+    data: Uint8Array | string | any, // eslint-disable-line @typescript-eslint/no-explicit-any
+    lifetimeMs?: number
+  ) => {
+    if (this.context.signalingClient.ready()) {
+      let uint8Data;
+      if (data instanceof Uint8Array) {
+        uint8Data = data;
+      } else if (typeof data === 'string') {
+        uint8Data = new TextEncoder().encode(data);
+      } else {
+        uint8Data = new TextEncoder().encode(JSON.stringify(data));
+      }
+      this.validateDataMessage(topic, uint8Data, lifetimeMs);
+      const message = SdkDataMessagePayload.create();
+      message.topic = topic;
+      message.lifetimeMs = lifetimeMs;
+      message.data = uint8Data;
+      const messageFrame = SdkDataMessageFrame.create();
+      messageFrame.dataMessagePayloads = [message];
+      this.context.signalingClient.sendDataMessage(messageFrame);
+    } else {
+      throw new Error('Signaling client is not ready');
```

Comment:
Who is catching these errors?

`realtimeSendDataMessage` will call an array of callbacks including this one. e.g. Assume that a builder has added their callbacks. If `sendDataMessageHandler` is the second element in an array, a builder's 3rd, 4th, and subsequent callbacks will not be invoked.

```
  realtimeSendDataMessage(
    topic: string,
    data: Uint8Array | string | any, // eslint-disable-line @typescript-eslint/no-explicit-any
    lifetimeMs?: number
  ): void {
    this.wrap(() => {
      for (const fn of this.state.sendDataMessageCallbacks) {
        fn(topic, data, lifetimeMs);
      }
    });
  }
```

Reply:
Error-handling here could be improved; we probably shouldn't throw here. We may want to buffer up the message to send later or return the message back to the sender indicating it wasn't delivered (similar to `throttled`).